drivers/spi/spi_sifive.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (c) 2018 SiFive Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SPI_LEVEL
+#include <logging/sys_log.h>
+
+#include "spi_context.h"
+#include "spi_sifive.h"
+
+#include <stdbool.h>
+
+#include <device.h>
+#include <spi.h>
+
+#define MIN(x, y) ((x < y) ? x : y)
+
+/* Structure Declarations */
+
+struct spi_sifive_data {
+	struct spi_context ctx;
+};
+
+struct spi_sifive_cfg {
+	u32_t base;
+	u32_t f_sys;
+};
+
+/* Helper Functions */
+
+int spi_config(struct device *dev, u32_t frequency, u16_t operation)
+{
+	if (SPI_OP_MODE_GET(operation) != SPI_OP_MODE_MASTER)
+		return -ENOTSUP;
+
+	if (operation & SPI_MODE_LOOP)
+		return -ENOTSUP;
+
+	/* Set the SPI frequency */
+	u32_t div = (SPI_CFG(dev)->f_sys / (2 * frequency)) - 1;
+
+	SPI_REG(dev, REG_SCKDIV) = (SF_SCKDIV_DIV_MASK & div);
+
+	/* Set the polarity */
+	if (operation & SPI_MODE_CPOL) {
+		/* If CPOL is set, then SCK idles at logical 1 */
+		SET_BIT(SPI_REG(dev, REG_SCKMODE), SF_SCKMODE_POL);
+	} else {
+		/* SCK idles at logical 0 */
+		CLR_BIT(SPI_REG(dev, REG_SCKMODE), SF_SCKMODE_POL);
+	}
+
+	/* Set the phase */
+	if (operation & SPI_MODE_CPHA) {
+		/*
+		 * If CPHA is set, then data is sampled
+		 * on the trailing SCK edge
+		 */
+		SET_BIT(SPI_REG(dev, REG_SCKMODE), SF_SCKMODE_PHA);
+	} else {
+		/* Data is sampled on the leading SCK edge */
+		CLR_BIT(SPI_REG(dev, REG_SCKMODE), SF_SCKMODE_PHA);
+	}
+
+	/* Get the frame length */
+	u32_t fmt_len = SPI_WORD_SIZE_GET(operation);
```

Comment:
do not declare a variable in the middle of a code block, always at the beginning. Apply that everywhere